### ISSUE TITLE
Unable to open swap file for "filename" on fresh new install

### ISSUE THREAD

**Issue by jeanleonino** [neovim/neovim#4530]
- Neovim version: 0.1.3-dev
- Operating system: Ubuntu 15.10
- Terminal emulator: zsh

### Actual behaviour
On a fresh new install the message `Unable to open swap file for "filename", recovery impossible` always show when opening any file. 

### Expected behaviour
This shouldn't be the standard behavior on a fresh new install. 

### Steps to reproduce
On a fresh new Ubuntu:

* install neovim
* open file
* observe

**Comment by jeanleonino** [comment]
This is not the same as #3898 - the problem is actually this being the behavior on a fresh new install.

**Comment by SaswatPadhi** [comment]
Bad permission for `swap` directory:

```
$ ll ~/.local/share/nvim
drwx------ 2 saswat saswat 4.0K Apr  5 23:26 shada/
drwxr-xr-x 3 saswat saswat 4.0K Apr  5 23:10 site/
drwxr-xr-x 2 root   root   4.0K Apr  5 19:55 swap/
drwxr-xr-x 2 saswat saswat 4.0K Apr  5 23:25 undo/
```

**Comment by jeanleonino** [comment]
@SaswatPadhi Yes, it is that. But this shouldn't happen since I installed the default way. Why is it happening?

**Comment by justinmk** [comment]
We should show a better error message for cases where we don't have write and/or read permissions to `'directory'`, `'undodir'`, etc.

**Comment by justinmk** [comment]
> But this shouldn't happen since I installed the default way. Why is it happening?

@jeanleonino What does `umask` say?

**Comment by jeanleonino** [comment]
>We should show a better error message for cases where we don't have write and/or read permissions to 'directory', 'undodir', etc.

@justinmk I don't like the `'undodir'` by default, it's easier to assume that the user is running the editor in a folder they have permission, so having by default a local swap file is more logical. 

> @jeanleonino What does umask say?

`002`

**Comment by ZyX-I** [comment]
@jeanleonino Choice of directories has nothing to do with permission assumptions. It is assumed that user does not want garbage in the file system. It is also logical to assume that directories under ~, especially ~/.local/… are writeable by the user: it is user home directory after all.

**Comment by jeanleonino** [comment]
>t is assumed that user does not want garbage in the file system. It is also logical to assume that directories under ~, especially ~/.local/… are writeable by the user: it is user home directory after all.

@ZyX-I well, here it says: https://github.com/neovim/neovim/blob/master/runtime/doc/undo.txt#L253 but the default install is not using it?

**Comment by ZyX-I** [comment]
@jeanleonino Unfortunately I did not know all the places where old default was mentioned in one or the other way. Actual default is at `:h 'undodir'`.

**Comment by ZyX-I** [comment]
#4559

**Comment by justinmk** [comment]
> I don't like the 'undodir' by default, it's easier to assume that the user is running the editor in a folder they have permission, so having by default a local swap file is more logical.

@jeanleonino We want to optimize for the common case. Not edge cases. Vast majority of users do not want swap files appearing next to each edited file.

**Comment by jeanleonino** [comment]
@ZyX-I it's ok. Yeah, there's this communication error #4559 solves. 

@justinmk But yet, this problem here is not solved. Because the default install on Ubuntu is not and edge case. And I am OK with sap files next to each edited file, that pattern can be confusing.  Why not the default folder be `~/.vim/undo/`?

**Comment by jeanleonino** [comment]
It's user specific, which is good, and it's also easier to handle permissions.

**Comment by justinmk** [comment]
> Because the default install on Ubuntu is not and edge case.

A possible bug in the ubuntu package, or our install, doesn't have anything to do with the general question of where to store the files by default. 

Other than your experience, what leads you to believe this affects all ubuntu neovim users?

> Why not the default folder be ~/.vim/undo/?

This is discussed at length in https://github.com/neovim/neovim/pull/3470 and it is documented.

**Comment by SaswatPadhi** [comment]
I can confirm that this is an issue with the general Ubuntu package; or with neovim install scripts, not just with @jeanleonino.

**Comment by jeanleonino** [comment]
@justinmk 
You can observe it in any fresh new Ubuntu installation. If you try to reproduce the bug.

**Comment by fwalch** [comment]
Reproduce how, exactly? If I do just `sudo apt-get install neovim; nvim some-file`, which is what is implied by your initial description, that error certainly doesn't occur. I can reproduce something similiar if I run `sudo nvim` as first command, which creates `.local/share/nvim` with root owner.

**Comment by SaswatPadhi** [comment]
@fwalch, I haven't tried on the latest build. But when I installed neovim on 15.10 and Ubuntu 16.04 (beta 2) with `sudo apt-get install neovim; nvim some-file` I had the error. 
I manually changed the permissions of `~/.local/share/nvim/swap` to fix this.

**Comment by kmike** [comment]
I just installed neovim on a fresh Ubuntu 16.04 (`sudo apt install neovim`); the issue still present - by default there is a warning.
